### ISSUE TITLE
[Chapter 18] Breakout DQN did not learn anything ?

### ISSUE THREAD

**Issue by lebaste77** [ageron/handson-ml2#397]
Hi,

I am facing somewhat something similar to issue #117 when executing the notebook. Here are the generated GIFs with different training lengths (n_iter = [10000, 20000, 22000]).
![breakout](https://user-images.githubusercontent.com/73541689/109432760-10cd8500-7a0d-11eb-95c2-a5311290aa2e.gif) ![breakout2](https://user-images.githubusercontent.com/73541689/109429283-4f5a4400-79fb-11eb-9a88-0862cc8a31de.gif) ![breakout](https://user-images.githubusercontent.com/73541689/109425743-23829280-79ea-11eb-8519-454613719c09.gif) 
The more I train it, the less it tries to win. Is there an inversion in the reward somewhere ?

I got the same notebook outputs as yours during training, and still got this behavior. The last output info (>22000 iterations) is :
```
INFO:absl: 
		 NumberOfEpisodes = 478
		 EnvironmentSteps = 86748
		 AverageReturn = 1.399999976158142
		 AverageEpisodeLength = 193.60000610351562
```
However, 193 steps does not seems to be the duration of my experience here. Do you have any idea ?
_PS : due to hardware constraints, I am using TF 2.3.0, tf-agents 0.6.0, tf-probability 0.11.1 ..._

----------------------

But it seems that I have a global bigger issue, since the CartPole model also does not behave as wished. It is learning a little bit, but seems to diverge. Here is an example with the Dueling  DQN. 
![dueling_dqn_CartPole](https://user-images.githubusercontent.com/73541689/109427671-c095f900-79f3-11eb-9fdf-274e6c34ec8a.gif)

Actually PG had better results in my case : 
![PG_CartPole](https://user-images.githubusercontent.com/73541689/109428508-a5c58380-79f7-11eb-8781-40e16efac016.gif)

Note that executing the CartPole in Colab generates exactly the same diverging behavior (below Dueling DQN) !
![dddqn_CartPole](https://user-images.githubusercontent.com/73541689/109430601-7cf6bb80-7a02-11eb-9e08-c3c3715ee4ca.gif)
(and Colab creates slower gifs)

--------------------

Finally, here is a strange behavior of the DQN, used with Colab with 20000 iterations of training... There is no ball (I retried it 3 times).
![breakout colab 20000](https://user-images.githubusercontent.com/73541689/109432613-4c1b8400-7a0c-11eb-9a22-0c35ae832aba.gif)

------------------
I checked if the problem was the gif export, but with a video it does the same behavior (see #117).

**Comment by ageron** [comment]
Hi @lebaste77 ,

Thanks for your detailed feedback.

I haven't changed the code in this notebook much since I wrote the book, apart minor updates when upgrading the libraries. Back then, training worked fine. I forgot exactly how long I trained the DQN for Breakout, but I think it was pretty long (at least hundreds of thousands of iterations), and it eventually reached this performance:
![breakout](https://github.com/ageron/handson-ml2/raw/master/images/rl/breakout.gif)
I remember that after just 10,000 steps it was really bad. So perhaps it's just a matter of training it for longer?

That said, I admit that it does look quite fishy. :/ One possible explanation is that I recently updated all the notebooks to work with TF 2.4, and I also updated TF Agents and Gym to the latest versions. Perhaps some default hyperparameters have changed? Or there's a new bug in TF Agents?

Regarding the cart pole, its performance is clearly lower than it used to be, but this might just be a normal random variation: the code is pure-python and it hasn't changed at all, so I don't really see where the drop in performance would come from if not for an unlucky initialization (TensorFlow's random number generator has changed since I wrote the book, so we're getting different random numbers than I got when writing the book). The performance does looks bad, since the cart pole seems to be falling towards the side of the screen, but if the fall is slow enough, it might actually be close to the maximum number of steps. If so, then as far as the algorithm is concerned, it might actually be close to optimal. So I suggest trying other random initializations to see if you can find one that works well (it is common for an RL algorithm to succeed less than 33% of the time depending on the random init).

If you manage to get good performance with different hyperparameters or random inits, I would really appreciate it if you could add a comment to this issue. 👍

**Comment by lebaste77** [comment]
Thank you for your answers.

Here is what I tried : 
**1. Breakout : Same init with 100000 iterations** 
IMO, something's wrong with my env here.. ? (yes, this is an animated gif file) 
![breakout_100000](https://user-images.githubusercontent.com/73541689/110203161-cadd4a80-7e6c-11eb-8c17-ef67eaf0ec85.gif)
https://user-images.githubusercontent.com/73541689/110203246-65d62480-7e6d-11eb-8cee-6a259d70ea2f.mp4
let's restart with another random init...

**2. Breakout : other random init with 20000 iterations**
`the_seed = 66`
![breakout_rand66](https://user-images.githubusercontent.com/73541689/110204725-e1879f80-7e74-11eb-9098-fd88f77ec633.gif)
https://user-images.githubusercontent.com/73541689/110204729-e4829000-7e74-11eb-833f-2a214b94cd09.mp4

**3. Breakout : other random init with ~235000 iterations (continuation of the previous)**
`the_seed = 66`
![breakout_rand66_235k_sansreset](https://user-images.githubusercontent.com/73541689/110212760-4bb43a80-7e9d-11eb-9c53-69ebedd4e8e6.gif)
https://user-images.githubusercontent.com/73541689/110212888-d301ae00-7e9d-11eb-9554-b2044e4a30aa.mp4
That's much better.
PS : for memory issue, I put the `replay_buffer `with `max_length=100000`

**4. Breakout : I also tried to add `tf_env.reset()` and `tf_env.pyenv.envs[0].step(np.array(1))` before running the `watch_driver`, I thought this could avoid having the env in a unknown state**
![breakout_rand66_235k_avecavecreset](https://user-images.githubusercontent.com/73541689/110212842-9fbf1f00-7e9d-11eb-9db3-5b7c7dffb6f1.gif)
https://user-images.githubusercontent.com/73541689/110212883-cc733680-7e9d-11eb-9388-3e9ad033fa60.mp4

Finally, it seems only to be a matter of init/nb of steps. (what a relief). Maybe for your example in notbook, you may suggest to do at least 200k iterations here ?

----------------
**Now for the CartPole issue : variying steps and episodes count, and random seeds**

```
the_seed = 42
n_episodes = 600
n_steps = 500
```
![dueling_dqn_CartPole_500e600](https://user-images.githubusercontent.com/73541689/110206166-6b873680-7e7c-11eb-9ca1-072540047b78.gif)
```
the_seed = 99
n_episodes = 1200
n_steps = 500
```
![dueling_dqn_CartPole_500e1200rdm99](https://user-images.githubusercontent.com/73541689/110206214-bb65fd80-7e7c-11eb-982b-dc6243312357.gif)

This shows that the objective has been fulfilled as the duration of those animations are respectively 290 and 341 steps.
In the previous animation with duelling DQN, the number of frames was only 161 in the animation.
NB : I used the seed value on all seeds : 
```
tf.random.set_seed(the_seed)
np.random.seed(the_seed)
env.seed(the_seed)
```

Thanks again for your answer, this may close this "issue".

**Comment by 8bitmp3** [comment]
> I remember that after just 10,000 steps it was really bad.

@ageron That's "very RL" 😄 Meta-RL to the rescue.

**Comment by ageron** [comment]
Thanks so much @lebaste77, this is really awesome work! 😃👍 
I'll update the notebook to use the seeds you found, and I'll add a note to explain that at least 200,000 iterations are needed for Breakout. This chapter was one of the hardest to write, and it took me forever to get the code to finally work. So I was afraid the problem was worst than that, perhaps a regression in one of the underlying libraries, that would have been a nightmare to fix. In short, you saved me hours of work, thanks again!!! 🎉

**Comment by ageron** [comment]
I've run some tests, and I've found two interesting things:
1. When using @lebaste77 's seed=99, the DQN actually works well **without any training at all**! 🤯 The random initialization of the neural net is just amazingly good. In fact, the training process actually breaks the model, so the performance goes down, not up. So I prefer not to use this seed, as I'd like to show a successful training process.
2. I tried to tweak other hyperparameters, and it turns out that just using a learning rate of 1e-2 instead of 1e-3 for the regular DQN results in a robust training process with a great result:
![image](https://user-images.githubusercontent.com/76661/110255054-c9d82480-7ff6-11eb-9117-a0a57270c526.png)
